subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/algorithm/cell.go

```diff
@@ -40,14 +41,34 @@ type Cell interface {
 	SetParent(Cell)
 	GetChildren() CellList
 	SetChildren(CellList)
+	AtOrHigherThanNode() bool
+
+	IncreaseFreeGpuNum(int32)
+	GetFreeGpuNumForPriority(CellPriority) int32
+	GetUsedGpuNumAtPriority(CellPriority) int32
+	IncreaseUsedGpuNumAtPriority(CellPriority, int32)
+	GetUsedGpuNumSamePriority() int32
+	GetUsedGpuNumOtherPriority() int32
+	SetUsedGpuNumForPriority(CellPriority)
+	SetUsedGpuNumAllPriority(CellPriority)
 }
 
 type GenericCell struct {
-	chain    CellChain
-	level    CellLevel
-	priority CellPriority // priority of a cell is max of those of its children and itself
-	parent   Cell         // pointer to its parent cell
-	children CellList     // pointer to its children cells
+	chain              CellChain
+	level              CellLevel
+	priority           CellPriority
+	parent             Cell     // pointer to its parent cell
+	children           CellList // pointer to its children cells
+	atOrHigherThanNode bool     // true if the cell is at or higher than node level
+
+	totalGpuNum           int32                  // total GPU number of a cell
+	freeGpuNum            int32                  // free GPU number of a cell
+	freeGpuNumForPriority int32                  // free GPU number for a certain priority (lower priority considered as free)
```

Comment:
>freeGpuNumForPriority [](start = 1, length = 21)

Align with usedGpuNumAtPriority, so rename to freeGpuNumAtPriority?

And why it is not a map, and what is difference with freeGpuNum?

Reply:
It is a single number because it is for a specific priority (input of GetFreeGpuNumAtPriority()).
freeGpuNum is the number of all free GPUs. freeGpuNumForPriority considers GPUs used by lower priorities also free.

---
In reply to: [332808421](https://github.com/microsoft/pai/pull/3684#discussion_r332808421) [](ancestors = 332808421)